# Basics of Python

## List of Topics

- [Functions](#functions)
- [Variables](#variables)
- [Data Types](#data-types)
- [Operators](#operators)
- [Control Flow](#control-flow)
- [Loops](#loops)
- [Strings](#strings)
- [Lists](#lists)
- [Tuples](#tuples)
- [Dictionaries](#dictionaries)
- [Sets](#sets)
- [Comprehensions](#comprehensions)
- [User Defined Functions](#user-defined-functions)
- [Lambda Functions](#lambda-functions)
- [Map](#map)
- [Filter](#filter)
- [Reduce](#reduce)
- [Modules](#modules)
    - [Built-in Modules](#built-in-modules)
        - [Math](#math)
        - [Random](#random)
        - [Datetime](#datetime)
        - [OS](#os)
        - [Sys](#sys)
        - [RE](#re)
- [Classes](#classes)

### Functions

Functions are blocks of code that will be executed whenever they are called. There are two types of functions in Python: built-in functions and user-defined functions.

In [6]:
print("Hello World!")

Hello World!


In [7]:
help(print)

Help on built-in function print in module builtins:

print(*args, sep=' ', end='\n', file=None, flush=False)
    Prints the values to a stream, or to sys.stdout by default.
    
    sep
      string inserted between values, default a space.
    end
      string appended after the last value, default a newline.
    file
      a file-like object (stream); defaults to the current sys.stdout.
    flush
      whether to forcibly flush the stream.



\n - Escape character or sequence or new line character

In [23]:
print("hi", "there", "how", "are", "you", sep=" ", end="; ")
print("hello world")

hi there how are you; hello world


In [3]:
print("Hello World")
print("Welcome")

Hello World
Welcome


### Variables

Variables are labels that point to a memory location that stores a value.

Why labels?

- Each byte of a memory has an address
- the addresses will be in hexadecimal format
- it is difficult to remember the addresses
- so, we use labels to point to the addresses

In [7]:
age = 20

In [10]:
print(age, type(age))

20 <class 'int'>


In [12]:
age = "twenty"

In [13]:
print(age, type(age))

twenty <class 'str'>


- How the codes are compiled and executed?

### Data Types

Datatypes tells us what type of data a variable is storing.

What are the different data types in python?

- int (integer) - 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
- float (floating point) - 3.45, 6.78, 9.01
- str (string) - "hello@#123", "world", "python"
- bool (boolean) - True, False (two types of boolean values)
- NoneType (None) - None (only one value)
- list - [1, 2, 3, 4, 5]
- tuple - (1, 2, 3, 4, 5) (read only list) (immutable meaning unchangeable or cannot be changed once created & assigned)
- dict (dictionary) - {"name": "python", "version": 3.8}
- set - set{1, 2, 3, 4, 5} (unique values) - mathematical set operations

In [14]:
percentage = 0.05
print(percentage, type(percentage))

0.05 <class 'float'>


Strings - Sequence of characters. For example: "Hello World", "Apple", "1234", "Apple@123"

In [16]:
sentence = "I am a string"
print(sentence, type(sentence))

I am a string <class 'str'>


In [15]:
name = 'apple'
print(name, type(name))

apple <class 'str'>


In [20]:
message = """This is a multi-line string
\n\tThis is the second line
this is the third line
"""
print(message, type(message))

This is a multi-line string

	This is the second line
this is the third line
 <class 'str'>


In [22]:
message='''This is a multi-line string\b'''
print(message)

This is a multi-line strin


Escape Sequences or Escape Characters

- \n - new line
- \t - tab
- \b - backspace
- \r - carriage return (moves the cursor to the beginning of the line)
- \f - form feed (moves the cursor to the next line)
- \v - vertical tab
- \a - alert (bell)
- \0 - null character

In [25]:
isValid = False

In [26]:
print(isValid, type(isValid))

False <class 'bool'>


In [29]:
name = None

In [30]:
print(name, type(name))

None <class 'NoneType'>


lists in python can store different types of data unlike arrays in other programming languages like C, C++, Java, etc where are arrays can store only one type of data.

In [32]:
# id, firstname, lastname, age, salary, isMarried, isWorking, creditScore, transactions
person101 = [101, "John", "Doe", 24, 10000, True, False, 735.5, [100, 200, 300]]

In [33]:
print(person101, type(person101))

[101, 'John', 'Doe', 24, 10000, True, False, 735.5, [100, 200, 300]] <class 'list'>


In [34]:
person102 = (102, "Jane", "Doe", 24, 10000, True, False, 735.5, [100, 200, 300])

In [35]:
print(person102, type(person102))

(102, 'Jane', 'Doe', 24, 10000, True, False, 735.5, [100, 200, 300]) <class 'tuple'>


In [36]:
d = {"name": "python", "version": 3.8}

In [37]:
print(d, type(d))

{'name': 'python', 'version': 3.8} <class 'dict'>


In [38]:
person101 = {"id": 101, "firstname": "John", "lastname": "Doe", "age": 24, "salary": 10000, "isMarried": True, "isWorking": False, "creditScore": 735.5, "transactions": [100, 200, 300]}

In [39]:
print(person101)

{'id': 101, 'firstname': 'John', 'lastname': 'Doe', 'age': 24, 'salary': 10000, 'isMarried': True, 'isWorking': False, 'creditScore': 735.5, 'transactions': [100, 200, 300]}


In [40]:
person101["transactions"]

[100, 200, 300]

In [41]:
a = set([1, 2, 3])
b = set([2, 3, 4])

a.union(b)

{1, 2, 3, 4}

Variables and DataTypes Exercises:

1. Create a variable called `name` and assign it a value of "John"
2. Create a variable called `age` and assign it a value of 20
3. Create a variable called `isNew` and assign it a value of True
4. Create a list called `names` and add 5 names to it
5. Let's say we have to store the details of a particular user who logs into github. Think of list of values you want to store for that user and create a list with those values as a dictionary.
6. Create a set called `numbers` and add 5 numbers to it.


In [42]:
# Exercise 1: Create a variable called name and assign it the value "John"
name = "John"
print(name)

John


In [44]:
age = 20
print(age)

20


In [45]:
isNew = True
print(isNew)

True


In [46]:
names = ["Inbarasan", "Santhosh", "Mohan", "Mohanalakshmi", "Sankari"]
print(names)

['Inbarasan', 'Santhosh', 'Mohan', 'Mohanalakshmi', 'Sankari']


In [ ]:
githubUser = {
    'username': 'githubscientist',
    'email': 'sathish.datascientist@gmail.com',
    'settings': {
        'publicProfile': {
            'name': 'Sathish',
            'publicEmail': 'sathish.datascientist@gmail.com',
            'bio': 'Data Scientist',
            'socialAccounts': ['https://linkedin.com/in/mailtosathish', 'https://twitter.com/sathish_twit']
        },
        'accountIsActive': True
    },
    'repositories': ['dsa', 'python', 'hello-world']
}

In [47]:
# Create a set called `numbers` and add 5 numbers to it.
numbers = set([1, 2, 3, 4, 5])
print(numbers, type(numbers))

{1, 2, 3, 4, 5} <class 'set'>


In [48]:
numbers = {1, 2, 3, 4, 5}
print(numbers, type(numbers))

{1, 2, 3, 4, 5} <class 'set'>


comment lines

In [ ]:
# the following code creates a variable name and assigns it the value "John"
name = "John"

In [ ]:
# single line comment

In [ ]:
'''
This is a multi-line comment
'''

In [ ]:
"""
This is a multi-line comment
"""

### Operators

Operator - A symbol that performs an operation on one or more operands.

Example:

5 + 6 -> expression

5, 6 -> operands
+ -> operator

Types of Operators:

- Arithmetic Operators
- Comparison Operators (Relational Operators)
- Logical Operators
- Bitwise Operators
- Assignment Operators

Arithmetic Operators:

+ addition
- subtraction
* multiplication
/ division
% modulus (remainder)
** exponentiation (power)
// floor division (quotient)

In [49]:
5 + 6

11

In [51]:
5 - 6

-1

In [52]:
5 * 6

30

In [53]:
5 / 6

0.8333333333333334

In [54]:
# floor division or gives the reminder
5 % 6

5

In [55]:
5 ** 3

125

In [56]:
5 ** 2

25

In [57]:
100 ** (1/2)

10.0

In [58]:
100 ** 0.5

10.0

In [59]:
5/6

0.8333333333333334

In [60]:
5/3

1.6666666666666667

In [61]:
5//3

1

Relational Operators:

== equal to
!= not equal to
> greater than
< less than
>= greater than or equal to
<= less than or equal to

The result of an expression that involves relational operators is a boolean value.

In [62]:
5 == 6

False

In [63]:
5 != 6

True

In [64]:
5 < 6

True

In [65]:
5 > 6

False

In [66]:
5 <= 6 # less than or equal to

True

In [67]:
5 <= 5

True

In [68]:
5 < 5

False

In [69]:
5 == 5

True

In [70]:
5 >= 5

True

In [72]:
age = 17
hasAadhar = True
hasVoterID = False

Eligibility Criteria:

- age should be greater than or equal to 18
- should have a valid ID proof
    - either aadhar card or voter id card

In [73]:
age >= 18

False

Logical Operators: they are helpful in combining multiple conditions or expressions.

and
or
not

In [75]:
age = 18
hasAadhar = True
hasVoterID = False

age >=18 and hasAadhar == True

True

logical 'and' table: only true if both the conditions are true otherwise false

|condition1 | condition2 | result = condition1 and condition2 |
|-----------|------------|------------------------------------|
|True       | True       | True                               |
|True       | False      | False                              |
|False      | True       | False                              |
|False      | False      | False                              |


logical 'or' table: true if any one of the conditions is true otherwise false

|condition1 | condition2 | result = condition1 or condition2 |
|-----------|------------|----------------------------------|
|True       | True       | True                             |
|True       | False      | True                             |
|False      | True       | True                             |
|False      | False      | False                            |

In [77]:
age = 18
hasAadhar = False
hasVoterID = True

age >= 18 and hasAadhar == True

False

In [78]:
age = 18
hasAadhar = False
hasVoterID = True

(age >= 18) and ((hasAadhar == True) or (hasVoterID == True))

True

In [82]:
not (6 > 6)

True

In [6]:
age = 17
hasAadhar = True
hasVoterID = True

(age >= 18 ) and ((hasAadhar == True) or (hasVoterID == True))

False

Operator Precedence

Bitwise Operators

& bitwise and
| bitwise or
^ bitwise xor
~ bitwise not or complement or negation
<< bitwise left shift
>> bitwise right shift

and truth table

|condition1 | condition2 | result = condition1 and condition2 |
|-----------|------------|------------------------------------|
|1          | 1          | 1                                  |
|1          | 0          | 0                                  |
|0          | 1          | 0                                  |
|0          | 0          | 0                                  |


In [7]:
5 & 6

4

In [8]:
13 & 23

5

In [10]:
15 & 27

11

In [11]:
5 | 6

7

or truth table

|condition1 | condition2 | result = condition1 or condition2 |
|-----------|------------|----------------------------------|
| 0         | 0          | 0                                |
| 0         | 1          | 1                                |
| 1         | 0          | 1                                |
| 1         | 1          | 1                                |


In [12]:
31 | 15

31

bitwise xor table

|condition1 | condition2 | result = condition1 xor condition2 |
|-----------|------------|-----------------------------------|
| 0         | 0          | 0                                 |
| 0         | 1          | 1                                 |
| 1         | 0          | 1                                 |
| 1         | 1          | 0                                 |

In [ ]:
5 ^ 6

Problem: Given a list of numbers, find the number that occurs only once. All other numbers occur twice.

Example:

Input: [1, 2, 3, 4, 5, 1, 2, 3, 4]

Algorithm 1:

Take a number from the list and check if it occurs again in the list.
If it occurs, change to the next number in the list.
If it doesn't occur, then it is the number we are looking for.

Algorithm 2:

Find the counts of all the numbers in the list.
Find the number that has a count of 1.

Algorithm 3: Cancelling out

Find the bitwise xor of all the numbers in the list.
The result will be the number that occurs only once.

Problem: Given a number, to check whether the number is even or odd.

1 - 0001
2 - 0010
3 - 0011
4 - 0100
5 - 0101
6 - 0110
7 - 0111
8 - 1000
9 - 1001
10 - 1010

Algorithm:

Take the number and check the last bit of the number.(do a bitwise and with 1). the result will be either 0 or 1.
If the result is 0, then the number is even.
If the result is 1, then the number is odd.

Question: How do we check whether the last bit is 0 or 1?

Solution: We can use the bitwise and operator.

Question: How do we check whether a particular bit is 0 or 1?

Solution: We can use the bitwise and operator. bitwise & with the bit position.

example: 

for checking the last bit, we can use bitwise & with 1.
for checking the second last bit, we can use bitwise & with 2.
for checking the third last bit, we can use bitwise & with 4.
for checking the fourth last bit, we can use bitwise & with 8.
for checking the fifth last bit, we can use bitwise & with 16.
...
...

Problem: To check whether a number is a power of 2 or not.

Example:

I/P: 2
O/P: Power of 2

I/P: 3
O/P: Not a power of 2

I/P: 4
O/P: Power of 2

I/P: 6
O/P: Not a power of 2

Algorithm:

Start from the first power of 2. i.e. 2^0 = 1
Generate all the next powers of 2 until the number is greater than or equal to the given number.
If the number is equal to the given number, then it is a power of 2.

Algorithm 2:

Take the number and check if the number & number - 1 is 0.
If it is 0, then it is a power of 2.
If it is not 0, then it is not a power of 2.

Example:

number = 4
number - 1 = 3

number & number - 1 = 4 & 3 = 0
As the result is 0, 4 is a power of 2.

number = 6

number - 1 = 5

number & number - 1 = 6 & 5 = 4
As the result is not 0, 6 is not a power of 2.

bitwise left shift operator (<<)

In [13]:
5 << 1

10

In [14]:
15 << 1

30

Problem: Given a number, find its multiple of 4, but without using the multiplication operator and without using the addition operator.

In [15]:
2<<1

4

In [16]:
3<<1

6

In [17]:
10<<1 # 2 multiple of the number

20

In [18]:
10 << 2

40

In [19]:
15 << 2 # 4 multiple of the number

60

In [20]:
5 << 3 # 8 multiple of the number

40

what if we want to find the m multiple of a number(n)?

In [22]:
number = 5
# find the 7 multiple of the number
(number<<3) - number

35

In [23]:
number = 5
# to find the 6 multiple of the number
# number * 6 
# without using multiplication operator
number<<3

30

In [24]:
5 >> 1

2

Problem: Given a decimal number, find the binary representation of the number.

Algorithm 1: LCM method

Take the number and divide it by 2.
Record the remainder and the quotient.
Repeat the process until the quotient is 0.
The binary representation of the number will be the remainders in reverse order.

Algorithm 2: Power of 2 method

Take the number and check which of the powers of 2 is less than or equal to the number.
Subtract the power of 2 from the number. Write 1 in the place of the power of 2.
Repeat the process until the number is 0.

Algorithm 3: right shift method

Do the following until the number is 0.
    - check the last bit of the number.
    - if the last bit is 0, then write 0.
    - if the last bit is 1, then write 1.
    - right shift the number by 1.

bitwise complement operator (~)

In [25]:
~5

-6

Step 1: Find the binary representation of the number.
Step 2: Flip all the bits in the binary representation.
Step 3: Now if the number is positive, then the number would have become negative.
and what we have is a negative binary representation of the number.
Step 4: To find the number, we need to follow the steps below.
    - Find the 2's complement of the number.
          - To find the 1's complement of the number, we need to flip all the bits in the binary representation of the number.
          - To add 1 to the 1's complement, we need to add 1 to the LSB of the 1's complement.
    - The result will be the number.